com.civicrm.hrjobroles/tests/phpunit/api/v3/HrJobRolesTest.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+require_once 'HrJobRolesTestBase.php';
+
+class api_v3_HrJobRolesTest extends HrJobRolesTestBase {
+
+
+  function setUp() {
+    parent::setUp();
+  }
+
+  function tearDown() {
+    parent::tearDown();
+  }
+
+  /**
+   * check with empty array
+   */
+  function testJobRoleCreateEmpty() {
+    $this->callAPIFailure('HrJobRoles', 'create', array());
+  }
+
+  /**
+   * check if required fields are not passed
+   */
+  function testJobRoleCreateWithoutJobContract() {
+    $params = array(
+      'title' => 'test role',
+    );
+    $this->callAPIFailure('HrJobRoles', 'create', $params);
+  }
+
+  /**
+   *  Test creating new role with valid parameters
+   */
+  function testJobRoleCreate() {
+    // create contact
+    $contactParams = array('first_name'=>'walter', 'last_name'=>'white');
+    $contactID = $this->individualCreate($contactParams);
+    // create contract
+    $contract = $this->createJobContract($contactID, date('Y-m-d', strtotime('-14 days')));
+
+    $params = ['job_contract_id' => $contract->id, 'title' => 'test role', 'sequential' => 1];
+    $result = $this->callAPISuccess('HrJobRoles', 'create', $params);
+    $this->assertEquals($result['values'][0]['job_contract_id'], $contract->id);
+    $this->assertEquals($result['values'][0]['title'], 'test role');
+  }
+
+  /**
+   *  Test creating new role with valid
+   * (location, region, department, level, cost center ) fields
+   *
+   */
+  function testJobRoleCreateWithValidOptionValues() {
+    // create contact
+    $contactParams = array('first_name'=>'walter', 'last_name'=>'white');
+    $contactID = $this->individualCreate($contactParams);
+    // create contract
+    $contract = $this->createJobContract($contactID, date('Y-m-d', strtotime('-14 days')));
+
+    // create option group and values
+    $this->createSampleOptionGroupsAndValues();
+
+    $params = [
+      'job_contract_id' => $contract->id,
+      'title' => 'test role',
+      'location' => 'amman',
+      'region' => 'south amman',
+      'department' => 'amman devs',
+      'level_type' => 'guru',
+      'cost_center' => 'abdali',
+      'sequential' => 1
+    ];
+    $result = $this->callAPISuccess('HrJobRoles', 'create', $params);
+    $this->assertEquals($result['values'][0]['location'], 'amman');
+    $this->assertEquals($result['values'][0]['region'], 'south amman');
+    $this->assertEquals($result['values'][0]['department'], 'amman devs');
+    $this->assertEquals($result['values'][0]['level_type'], 'guru');
+    $this->assertEquals($result['values'][0]['cost_center'], 'abdali');
+  }
+
+  /**
+   *  Test creating new role with Invalid
+   * (location, region, department, level, cost center ) fields
+   *
+   */
+  function testJobRoleCreateWithInvalidOptionValues() {
+    // create contact
```

Comment:
If someone removes a pseudoconstant from the schema/DAO, any value should be accepted, so, if that happens, this test will need to be removed as well. I can't see why the code will break because of that, unless you are expecting a specific value for that field (but then the test would be different and not depend on the API).

In fact, as you said yourself, if you use the create() method you do can create a job role with an invalid option value. If that is really a problem (I don't think it is) we should not allow that to happen and fix the create() method!

What concerns me here is that this is not testing a single line of code from civihr. That is, the test is more likely to  break because of a change in how this validation thing in the api works (I hate this civicrm API!), than in a change in our code. You didn't added any specific code to the job roles API, so there's nothing to test there.

Now, if you really want to make sure that the pseudoconstants exists (I don't think that this is necessary), a better way would be to call the fields() method in the BAO and assert that the pseudoconstant is in the field definition. And also change the test name to clearly state what you're trying to test.

I think that you're afraid the the problem introduced by adding the pseudoconstants will happen again if we don't keep this test, right? This type of test won't help you with that. If the problem was on the import feature, that is what needs to be tested. We also had a problem creating the job roles using the UI, right? So we should also check if the javascript tests need to be fixed to cover any missing scenario.